src/data/loaders.py

```diff
@@ -1,113 +1,423 @@
+"""A module for lyrics and MIDI dataset loaders.
+
+There are two loaders: MIDILoader and LyricsLoader. These loaders are
+essentially file readers which additionally turn the data into a series of
+integer tokens.
+"""
 import os
 import numpy as np
 import logging
+import nltk
+import string
+import collections
+import math
+from operator import itemgetter
+import pretty_midi
 
-log = logging.getLogger("few-shot")
+_SUSTAIN_ON = 0
+_SUSTAIN_OFF = 1
+_NOTE_ON = 2
+_NOTE_OFF = 3
 
+NOTE_ON = 1
+NOTE_OFF = 2
+TIME_SHIFT = 3
+VELOCITY = 4
 
-def tokenize_lyrics_file(filepath, max_len):
-    """Turns a file into a list of "words"
+MAX_SHIFT_STEPS = 100
 
-    Arguments:
-        filepath (str): path to the lyrics file. e.g.
-            "/home/user/lyrics_data/tool/lateralus.txt"
-        max_len (int): the maximum length
+MIN_MIDI_VELOCITY = 1
+MAX_MIDI_VELOCITY = 127
+
+MIN_MIDI_PROGRAM = 0
+MAX_MIDI_PROGRAM = 127
+PROGRAMS_PER_FAMILY = 8
+
+
+log = logging.getLogger("few-shot")
+
+
+class Loader(object):
+    """A class for turning data into a sequence of tokens.
     """
-    token_count = 0
-    for line in open(filepath, 'r', errors='ignore'):
-        for token in line.split():
-            yield token
-            token_count += 1
-            if token_count >= max_len:
-                return
+    def __init__(self, max_len, dtype=np.int32, persist=True):
+        self.max_len = max_len
+        self.dtype = dtype
+        self.persist = persist
 
+    def is_song(self, filepath):
+        raise NotImplementedError
 
-def tokenize_midi_file(filepath):
-    """Turns a MIDI file into a list of event IDs.
+    def read(self, filepath):
+        raise NotImplementedError
 
-    Arguments:
-        filepath (str): path to the lyrics file. e.g.
-            "/home/user/freemidi_data/Tool/lateralus.mid"
+    def tokenize(self, data):
+        raise NotImplementedError
+
+    def detokenize(self, numpy_data):
+        raise NotImplementedError
+
+    def validate(self, filepath):
+        try:
+            self.load(filepath)
+            return True
+        except OSError:
+            return False
+        except KeyError:
+            return False
+        except EOFError:
+            return False
+        except IndexError:
+            return False
+        except ValueError:
+            return False
+
+    def load(self, filepath):
+        npfile = '%s.%s.npy' % (filepath, self.max_len)
+        if self.persist and os.path.isfile(npfile):
+            return np.load(npfile).astype(self.dtype)
+        else:
+            data = self.read(filepath)
+            tokens = self.tokenize(data)
+            numpy_tokens = np.zeros(self.max_len, dtype=self.dtype)
+            for token_index in range(min(self.max_len, len(tokens))):
+                numpy_tokens[token_index] = tokens[token_index]
+            if self.persist:
+                np.save(npfile, numpy_tokens)
+            return numpy_tokens
+
+
+class MIDILoader(Loader):
```

Comment:
this file is getting fairly large. how about creating a `loaders` directory, with `base_loader.py` holding `Loader`, `midi_loader.py` holding `MidiLoader`, etc.? this will also make it cleaner if we add unit tests (which are always a good idea :) ).

Reply:
Sure.